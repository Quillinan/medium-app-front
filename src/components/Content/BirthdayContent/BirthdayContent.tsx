import { useEffect, useState, useCallback } from 'react';
import { Birthday, ErrorResponse } from '../../../utils/types';
import { getMonthlyBirthdays } from '../../../services/getMonthlyBirthdays';
import { showLoading } from '../../../utils/loadingHelper';
import TableBirthday from '../../TableBirthday/TableBirthday';

const BirthdayContent: React.FC = () => {
  const [data, setData] = useState<Birthday[]>([]);
  const [loading, setLoading] = useState<boolean>(true);
  const [error, setError] = useState<string | null>(null);

  const sortData = useCallback((birthdays: Birthday[]) => {
    return birthdays.sort((a, b) => a.DTNASCIMENTO - b.DTNASCIMENTO);
  }, []);

  useEffect(() => {
    const fetchData = async () => {
      try {
        const response = await showLoading(
          getMonthlyBirthdays(new Date().getMonth() + 1)
        );

        if (Array.isArray(response)) {
          const sortedData = sortData(response);
          setData(sortedData);
        } else {
          const errorResponse = response as ErrorResponse;
          setError(errorResponse.message || 'Erro desconhecido');
        }
      } catch (error) {
        setError('Erro ao carregar os dados da API');
        console.error(error);
      } finally {
        setLoading(false);
      }
    };

    fetchData();
  }, [sortData]);

  return (
    <div className='flex items-center justify-center h-full'>
      {loading ? (
        <p>Carregando...</p>
      ) : error ? (
        <p>{error}</p>
      ) : (
        <TableBirthday data={data} />
      )}
    </div>
  );
};

export default BirthdayContent;
