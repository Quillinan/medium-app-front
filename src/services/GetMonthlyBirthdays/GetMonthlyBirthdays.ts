import { apiUrl, codeAffiliate, codeSystem, get } from '@services/Api/Api';
import { Birthday, ErrorResponse } from '@utils/types';

export const getMonthlyBirthdays = async (
  month: number
): Promise<Birthday[] | ErrorResponse | undefined> => {
  const codSentence = import.meta.env.VITE_API_SENTENCE_BIRTHDAYS;
  const url = `${apiUrl}/${codSentence}/${codeAffiliate}/${codeSystem}?parameters=MES_S%3D${month}`;

  const response = await get(url);

  return response as Birthday[];
};
