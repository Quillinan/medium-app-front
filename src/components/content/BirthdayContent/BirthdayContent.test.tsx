import { render, screen, waitFor } from '@testing-library/react';
import BirthdayContent from './BirthdayContent';
import { getMonthlyBirthdays } from '@services/GetMonthlyBirthdays/GetMonthlyBirthdays';
import { showLoading } from '@utils/LoadingHelper/LoadingHelper';

vi.mock('@services/GetMonthlyBirthdays/getMonthlyBirthdays', () => ({
  getMonthlyBirthdays: vi.fn(),
}));
vi.mock('@utils/LoadingHelper/LoadingHelper', () => ({
  showLoading: vi.fn(),
}));

describe('BirthdayContent', () => {
  const mockBirthdays = [
    { NOME: 'Alice', DTNASCIMENTO: 1 },
    { NOME: 'Bob', DTNASCIMENTO: 2 },
  ];

  it('should display custom error message from API response', async () => {
    const apiErrorResponse = { message: 'Erro no servidor' };
    (getMonthlyBirthdays as jest.Mock).mockResolvedValue(apiErrorResponse);
    (showLoading as jest.Mock).mockResolvedValue(apiErrorResponse);

    render(<BirthdayContent />);

    expect(showLoading).toHaveBeenCalled();
    const error = await screen.findByText('Erro no servidor');
    expect(error).toBeInTheDocument();
  });

  it('should display error message if fetching data fails', async () => {
    const errorMessage = 'Erro ao carregar os dados da API';
    (getMonthlyBirthdays as jest.Mock).mockRejectedValue(
      new Error(errorMessage)
    );
    (showLoading as jest.Mock).mockRejectedValue(new Error(errorMessage));

    render(<BirthdayContent />);

    expect(showLoading).toHaveBeenCalled();
    const error = await screen.findByText(errorMessage);
    expect(error).toBeInTheDocument();
  });

  it('should display birthday data when fetching is successful', async () => {
    (getMonthlyBirthdays as jest.Mock).mockResolvedValue(mockBirthdays);
    (showLoading as jest.Mock).mockResolvedValue(mockBirthdays);

    render(<BirthdayContent />);

    expect(showLoading).toHaveBeenCalled();

    await waitFor(() => {
      expect(screen.getByText('Alice')).toBeInTheDocument();
      expect(screen.getByText('Bob')).toBeInTheDocument();
    });
  });

  it('should sort the birthdays by date', async () => {
    (getMonthlyBirthdays as jest.Mock).mockResolvedValue(mockBirthdays);
    (showLoading as jest.Mock).mockResolvedValue(mockBirthdays);

    render(<BirthdayContent />);

    expect(showLoading).toHaveBeenCalled();

    await waitFor(() => {
      const aliceName = screen.getByTestId('name-Alice');
      const bobName = screen.getByTestId('name-Bob');

      expect(aliceName).toBeInTheDocument();
      expect(bobName).toBeInTheDocument();

      const names = [aliceName.textContent, bobName.textContent];
      expect(names).toEqual(['Alice', 'Bob']);
    });
  });

  it('should match snapshot', () => {
    const { asFragment } = render(<BirthdayContent />);
    expect(asFragment()).toMatchSnapshot();
  });
});
