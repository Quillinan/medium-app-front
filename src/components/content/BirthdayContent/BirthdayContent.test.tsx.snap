// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`BirthdayContent > should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="flex items-center justify-center h-full"
    data-testid="birthday-content"
  >
    <div
      class="overflow-x-auto"
    >
      <table
        class="min-w-full bg-white border border-gray-400 rounded-lg shadow-md"
      >
        <thead>
          <tr>
            <th
              class="px-4 py-2 border-b-2 border-gray-400 bg-gray-200 text-left text-sm font-semibold text-gray-700 uppercase tracking-wider"
            >
              Nome
            </th>
            <th
              class="px-4 py-2 border-b-2 border-gray-400 bg-gray-200 text-center text-sm font-semibold text-gray-700 uppercase tracking-wider"
            >
              Dia
            </th>
          </tr>
        </thead>
        <tbody>
          <tr>
            <td
              class="px-4 py-2 text-center text-sm text-gray-500"
              colspan="2"
            >
              Nenhum aniversariante encontrado.
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</DocumentFragment>
`;
